File: src/factories/MetadataCommentTagFactory.ts
import ts from "typescript";

import { IMetadataTypeTag } from "../schemas/metadata/IMetadataTypeTag";
import { Metadata } from "../schemas/metadata/Metadata";

import { Writable } from "../typings/Writable";

import { FormatCheatSheet } from "../tags/internal/FormatCheatSheet";
import { MetadataFactory } from "./MetadataFactory";
import { MetadataTypeTagFactory } from "./MetadataTypeTagFactory";

/**
 * @internal
 */
export namespace MetadataCommentTagFactory {
  export const analyze = (props: {
    errors: MetadataFactory.IError[];
    metadata: Metadata;
    tags: ts.JSDocTagInfo[];
    explore: MetadataFactory.IExplore;
  }): void => {
    // PREPARE MESSAGE CONTAINER
    const messages: string[] = [];
    const report = (msg: string) => {
      messages.push(msg);
      return null;
    };
    const validateReport = (next: {
      property: string | null;
      message: string;
    }): false => {
      messages.push(
        `the property ${
          next.property === null
            ? `["typia.tag"]`
            : `["typia.tag.${next.property}"]`
        } ${next.message}.`,
      );
      return false;
    };

    // VALIDATE AND CONSTRUCT COMMENT TAGS
    for (const tag of props.tags) {
      const tagger: TagRecord | null = parse({
        report,
        tag,
      });
      if (tagger === null) continue;
      for (const [key, value] of Object.entries(tagger)) {
        const filtered: IMetadataTypeTag[] = value.filter(
          (v) => v.validate !== null,
        ) as IMetadataTypeTag[];
        if (key === "array") {
          if (props.metadata.arrays.length === 0) {
            report(`requires array type`);
            continue;
          }
          for (const a of props.metadata.arrays) {
            Writable(a).tags = a.tags.filter((x) =>
              MetadataTypeTagFactory.validate({
                report: validateReport,
                type: "array",
                tags: [...x, ...filtered],
              }),
            );
            if (a.tags.length === 0) a.tags.push(filtered);
            else for (const tags of a.tags) tags.push(...filtered);
          }
        } else {
          const atomic = props.metadata.atomics.find((a) => a.type == key);
          if (atomic === undefined)
            if (key === "bigint" || key === "number") {
              const opposite = key === "bigint" ? "number" : "bigint";
              if (
                tagger[opposite] !== undefined &&
                props.metadata.atomics.some((a) => a.type === opposite)
              )
                continue;
            } else if (
              key === "string" &&
              value[0]?.kind === "format" &&
              value[0]?.value === "date-time"
            )
              continue;
            else report(`requires ${key} type`);
          else {
            Writable(atomic).tags = atomic.tags.filter((x) =>
              MetadataTypeTagFactory.validate({
                report: validateReport,
                type: key as "string",
                tags: [...x, ...filtered],
              }),
            );
            if (atomic.tags.length === 0) atomic.tags.push(filtered);
            else for (const tags of atomic.tags) tags.push(...filtered);
          }
        }
      }
    }

    // DO REPORT
    if (messages.length !== 0)
      props.errors.push({
        name: "comment tag(s)",
        explore: props.explore,
        messages,
      });
  };

  const parse = (props: {
    report: (msg: string) => null;
    tag: ts.JSDocTagInfo;
  }): TagRecord | null => {
    const next = PARSER[props.tag.name];
    if (next === undefined) return {};

    const value = (props.tag.text || [])[0]?.text;
    if (value === undefined && props.tag.name !== "uniqueItems")
      return props.report(`no comment tag value`);
    return next({
      report: props.report,
      value: value!,
    });
  };

  export const get = (props: {
    kind: string;
    type: "array" | "bigint" | "number" | "string";
    value: string;
  }): IMetadataTypeTag[] => {
    const output: IMetadataTypeTag[] | undefined = PARSER[props.kind]?.({
      report: () => null,
      value: props.value,
    })?.[props.type];
    if (output === undefined)
      throw new Error(
        `no tag found for (kind: ${props.kind}, type: ${props.type}).`,
      );
    return output;
  };
}

/**
 * @internal
 */
type TagRecord = {
  [P in Target]?: NotDeterminedTypeTag[];
};

/**
 * @internal
 */
type Target = "bigint" | "number" | "string" | "array";

/**
 * @internal
 */
type NotDeterminedTypeTag = Omit<IMetadataTypeTag, "validate" | "schema"> & {
  validate: string | undefined;
  schema: object | undefined;
};

/**
 * @internal
 */
const PARSER: Record<
  string,
  (props: { report: (msg: string) => null; value: string }) => {
    [P in Target]?: NotDeterminedTypeTag[];
  }
> = {
  /* -----------------------------------------------------------
        ARRAY
    ----------------------------------------------------------- */
  items: ({ report, value }) => ({
    array: [
      {
        name: `MinItems<${value}>`,
        target: "array",
        kind: "minItems",
        value: parse_integer({
          report,
          value,
          unsigned: true,
        }),
        validate: `${value} <= $input.length`,
        exclusive: true,
        schema: {
          minItems: parse_integer({
            report,
            value,
            unsigned: true,
          }),
        },
      },
      {
        name: `MaxItems<${value}>`,
        target: "array",
        kind: "maxItems",
        value: parse_integer({
          report,
          value,
          unsigned: true,
        }),
        validate: `$input.length <= ${value}`,
        exclusive: true,
        schema: {
          maxItems: parse_integer({
            report,
            unsigned: true,
            value,
          }),
        },
      },
    ],
  }),
  minItems: ({ report, value }) => ({
    array: [
      {
        name: `MinItems<${value}>`,
        target: "array",
        kind: "minItems",
        value: parse_integer({
          report,
          value,
          unsigned: true,
        }),
        validate: `${value} <= $input.length`,
        exclusive: true,
        schema: {
          minItems: parse_integer({
            report,
            value,
            unsigned: true,
          }),
        },
      },
    ],
  }),
  maxItems: ({ report, value }) => ({
    array: [
      {
        name: `MaxItems<${value}>`,
        target: "array",
        kind: "maxItems",
        value: parse_integer({
          report,
          value,
          unsigned: true,
        }),
        validate: `$input.length <= ${value}`,
        exclusive: true,
        schema: {
          maxItems: parse_integer({
            report,
            value,
            unsigned: true,
          }),
        },
      },
    ],
  }),
  uniqueItems: () => ({
    array: [
      {
        name: `UniqueItems`,
        target: "array",
        kind: "uniqueItems",
        value: true,
        validate: `$input.length <= 1 || (new Set($input).size === $input.length)`,
        exclusive: true,
        schema: {
          uniqueItems: true,
        },
      },
    ],
  }),

  /* -----------------------------------------------------------
        NUMBER
    ----------------------------------------------------------- */
  type: ({ value }) => {
    // EMENDATIONS
    if (value.startsWith("{") && value.endsWith("}"))
      value = value.substring(1, value.length - 1);
    if (value === "int") value = "int32";
    else if (value === "uint") value = "uint32";

    // MUST BE ONE OF THEM
    if (
      ["int32", "uint32", "int64", "uint64", "float", "double"].includes(
        value,
      ) === false
    )
      return {};
    return {
      number: [
        {
          name: `Type<${JSON.stringify(value)}>`,
          target: "number",
          kind: "type",
          value: value,
          validate:
            value === "int32"
              ? `Math.floor($input) === $input && -2147483648 <= $input && $input <= 2147483647`
              : value === "uint32"
                ? `Math.floor($input) === $input && 0 <= $input && $input <= 4294967295`
                : value === "int64"
                  ? `Math.floor($input) === $input && -9223372036854775808 <= $input && $input <= 9223372036854775807`
                  : value === "uint64"
                    ? `Math.floor($input) === $input && 0 <= $input && $input <= 18446744073709551615`
                    : value === "float"
                      ? `-1.175494351e38 <= $input && $input <= 3.4028235e38`
                      : `true`,
          exclusive: true,
          schema: ["int32", "int64"].includes(value)
            ? { type: "integer" }
            : ["uint32", "uint64"].includes(value)
              ? { type: "integer", minimum: 0 }
              : undefined,
        },
      ],
      bigint:
        value === "int64" || "uint64"
          ? [
              {
                name: `Type<${JSON.stringify(value)}>`,
                target: "bigint",
                kind: "type",
                value: value,
                validate: value === "int64" ? "true" : "BigInt(0) <= $input",
                exclusive: true,
                schema: value === "uint64" ? { minimum: 0 } : undefined,
              },
            ]
          : [],
    };
  },
  minimum: (props) => ({
    number: [
      {
        name: `Minimum<${props.value}>`,
        target: "number",
        kind: "minimum",
        value: parse_number(props),
        validate: `${props.value} <= $input`,
        exclusive: ["minimum", "exclusiveMinimum"],
        schema: {
          minimum: parse_number(props),
        },
      },
    ],
    bigint: [
      {
        name: `Minimum<${props.value}n>`,
        target: "bigint",
        kind: "minimum",
        value: (() => {
          const parsed = parse_integer({
            report: props.report,
            value: props.value,
            unsigned: false,
          });
          return parsed === null ? null : BigInt(parsed);
        })(),
        validate: `${props.value} <= $input`,
        exclusive: ["minimum", "exclusiveMinimum"],
        schema: {
          minimum: parse_number(props),
        },
      },
    ],
  }),
  maximum: (props) => ({
    number: [
      {
        name: `Maximum<${props.value}>`,
        target: "number",
        kind: "maximum",
        value: parse_number(props),
        validate: `$input <= ${props.value}`,
        exclusive: ["maximum", "exclusiveMaximum"],
        schema: {
          maximum: parse_number(props),
        },
      },
    ],
    bigint: [
      {
        name: `Maximum<${props.value}n>`,
        target: "bigint",
        kind: "maximum",
        value: (() => {
          const parsed = parse_integer({
            report: props.report,
            value: props.value,
            unsigned: false,
          });
          return parsed === null ? null : BigInt(parsed);
        })(),
        validate: `$input <= ${props.value}`,
        exclusive: ["maximum", "exclusiveMaximum"],
        schema: {
          maximum: parse_number(props),
        },
      },
    ],
  }),
  exclusiveMinimum: (props) => ({
    number: [
      {
        name: `ExclusiveMinimum<${props.value}>`,
        target: "number",
        kind: "exclusiveMinimum",
        value: parse_number(props),
        validate: `${props.value} < $input`,
        exclusive: ["minimum", "exclusiveMinimum"],
        schema: {
          exclusiveMinimum: parse_number(props),
        },
      },
    ],
    bigint: [
      {
        name: `ExclusiveMinimum<${props.value}n>`,
        target: "bigint",
        kind: "exclusiveMinimum",
        value: (() => {
          const parsed = parse_integer({
            report: props.report,
            value: props.value,
            unsigned: false,
          });
          return parsed === null ? null : BigInt(parsed);
        })(),
        validate: `${props.value} < $input`,
        exclusive: ["minimum", "exclusiveMinimum"],
        schema: {
          exclusiveMinimum: parse_number(props),
        },
      },
    ],
  }),
  exclusiveMaximum: (props) => ({
    number: [
      {
        name: `ExclusiveMaximum<${props.value}>`,
        target: "number",
        kind: "exclusiveMaximum",
        value: parse_number(props),
        validate: `$input < ${props.value}`,
        exclusive: ["maximum", "exclusiveMaximum"],
        schema: {
          exclusiveMaximum: parse_number(props),
        },
      },
    ],
    bigint: [
      {
        name: `ExclusiveMaximum<${props.value}n>`,
        target: "bigint",
        kind: "exclusiveMaximum",
        value: (() => {
          const parsed = parse_integer({
            report: props.report,
            value: props.value,
            unsigned: false,
          });
          return parsed === null ? null : BigInt(parsed);
        })(),
        validate: `$input < ${props.value}`,
        exclusive: ["maximum", "exclusiveMaximum"],
        schema: {
          exclusiveMaximum: parse_number(props),
        },
      },
    ],
  }),
  multipleOf: (props) => ({
    number: [
      {
        name: `MultipleOf<${props.value}>`,
        target: "number",
        kind: "multipleOf",
        value: parse_number(props),
        validate: `$input % ${props.value} === 0`,
        exclusive: true,
        schema: {
          multipleOf: parse_number(props),
        },
      },
    ],
    bigint: [
      {
        name: `MultipleOf<${props.value}n>`,
        target: "bigint",
        kind: "multipleOf",
        value: (() => {
          const parsed = parse_integer({
            report: props.report,
            value: props.value,
            unsigned: false,
          });
          return parsed === null ? null : BigInt(parsed);
        })(),
        validate: `$input % ${props.value}n === 0n`,
        exclusive: true,
        schema: {
          multipleOf: parse_number(props),
        },
      },
    ],
  }),

  /* -----------------------------------------------------------
        STRING
    ----------------------------------------------------------- */
  format: ({ value }) => {
    const matched = FORMATS.get(value);
    if (matched === undefined) return {};
    return {
      string: [
        {
          name: `Format<${JSON.stringify(matched[0])}>`,
          target: "string",
          kind: "format",
          value: matched[0],
          validate: matched[1],
          exclusive: true,
          schema: {
            format: matched[0],
          },
        },
      ],
    };
  },
  pattern: ({ value }) => ({
    string: [
      {
        name: `Pattern<${JSON.stringify(value)}>`,
        target: "string",
        kind: "pattern",
        value: value,
        validate: `RegExp(${JSON.stringify(value)}).test($input)`,
        exclusive: ["format"],
        schema: {
          pattern: value,
        },
      },
    ],
  }),
  length: (props) => ({
    string: [
      {
        name: `MinLength<${props.value}>`,
        target: "string",
        kind: "minLength",
        value: parse_number(props),
        validate: `${props.value} <= $input.length`,
        exclusive: true,
        schema: {
          minLength: parse_number(props),
        },
      },
      {
        name: `MaxLength<${props.value}>`,
        target: "string",
        kind: "maxLength",
        value: parse_number(props),
        validate: `$input.length <= ${props.value}`,
        exclusive: true,
        schema: {
          maxLength: parse_number(props),
        },
      },
    ],
  }),
  minLength: ({ report, value }) => ({
    string: [
      {
        name: `MinLength<${value}>`,
        target: "string",
        kind: "minLength",
        value: parse_number({ report, value }),
        validate: `${value} <= $input.length`,
        exclusive: true,
        schema: {
          minLength: parse_number({ report, value }),
        },
      },
    ],
  }),
  maxLength: ({ report, value }) => ({
    string: [
      {
        name: `MaxLength<${value}>`,
        target: "string",
        kind: "maxLength",
        value: parse_number({ report, value }),
        validate: `$input.length <= ${value}`,
        exclusive: true,
        schema: {
          maxLength: parse_number({ report, value }),
        },
      },
    ],
  }),
};

/**
 * @internal
 */
const parse_number = (props: {
  report: (msg: string) => null;
  value: string;
}): number | null => {
  const parsed: number = Number(props.value);
  if (isNaN(parsed) === true) return props.report(`invalid number`);
  return parsed;
};

/**
 * @internal
 */
const parse_integer = (props: {
  report: (msg: string) => null;
  unsigned: boolean;
  value: string;
}): number | null => {
  const parsed: number | null = parse_number(props);
  if (parsed === null) return null;
  else if (Math.floor(parsed) !== parsed)
    return props.report(`invalid integer`);
  else if (props.unsigned === true && parsed < 0)
    return props.report(`invalid unsigned integer`);
  return parsed;
};

/**
 * @internal
 */
const FORMATS: Map<string, [string, string]> = new Map([
  ...Object.entries(FormatCheatSheet).map(
    ([key, value]) => [key, [key, value]] as any,
  ),
  ["datetime", ["date-time", `!isNaN(new Date($input).getTime())`]],
  ["dateTime", ["date-time", `!isNaN(new Date($input).getTime())`]],
]);

File: packages/typia/native/core/factories/MetadataCommentTagFactory.go
package factories

import (
	"math"
	"strconv"
	"strings"

	schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type metadataCommentTagFactoryNamespace struct{}

var MetadataCommentTagFactory = metadataCommentTagFactoryNamespace{}

type metadataCommentTagFactory_TagRecord map[string][]schemametadata.IMetadataTypeTag

type metadataCommentTagFactory_parser func(props struct {
	Report func(msg string) any
	Value  string
}) metadataCommentTagFactory_TagRecord

func (metadataCommentTagFactoryNamespace) Analyze(props struct {
	Errors   *[]MetadataFactory_IError
	Metadata *schemametadata.MetadataSchema
	Tags     []schemametadata.IJsDocTagInfo
	Explore  MetadataFactory_IExplore
}) {
	messages := []string{}
	report := func(msg string) any {
		messages = append(messages, msg)
		return nil
	}
	validateReport := func(next struct {
		Property *string
		Message  string
	}) bool {
		property := "[\"typia.tag\"]"
		if next.Property != nil {
			property = "[\"typia.tag." + *next.Property + "\"]"
		}
		messages = append(messages, "the property "+property+" "+next.Message+".")
		return false
	}

	for _, tag := range props.Tags {
		tagger := metadataCommentTagFactory_parse(struct {
			Report func(msg string) any
			Tag    schemametadata.IJsDocTagInfo
		}{Report: report, Tag: tag})
		if tagger == nil {
			continue
		}
		for key, value := range tagger {
			filtered := []schemametadata.IMetadataTypeTag{}
			for _, elem := range value {
				filtered = append(filtered, elem)
			}
			if key == "array" {
				if len(props.Metadata.Arrays) == 0 {
					report("requires array type")
					continue
				}
				for _, array := range props.Metadata.Arrays {
					rows := [][]schemametadata.IMetadataTypeTag{}
					for _, row := range array.Tags {
						candidate := append(append([]schemametadata.IMetadataTypeTag{}, row...), filtered...)
						if MetadataTypeTagFactory.Validate(struct {
							Report func(struct {
								Property *string
								Message  string
							}) bool
							Type string
							Tags []schemametadata.IMetadataTypeTag
						}{Report: validateReport, Type: "array", Tags: candidate}) {
							rows = append(rows, row)
						}
					}
					array.Tags = rows
					if len(array.Tags) == 0 {
						array.Tags = append(array.Tags, filtered)
					} else {
						for i := range array.Tags {
							array.Tags[i] = append(array.Tags[i], filtered...)
						}
					}
				}
				continue
			}

			var atomic *schemametadata.MetadataAtomic
			for _, candidate := range props.Metadata.Atomics {
				if candidate.Type == key {
					atomic = candidate
					break
				}
			}
			if atomic == nil {
				if key == "bigint" || key == "number" {
					opposite := "bigint"
					if key == "bigint" {
						opposite = "number"
					}
					if _, ok := tagger[opposite]; ok {
						found := false
						for _, a := range props.Metadata.Atomics {
							if a.Type == opposite {
								found = true
								break
							}
						}
						if found {
							continue
						}
					}
				} else if key == "string" && len(value) != 0 && value[0].Kind == "format" && value[0].Value == "date-time" {
					continue
				}
				report("requires " + key + " type")
				continue
			}
			rows := [][]schemametadata.IMetadataTypeTag{}
			for _, row := range atomic.Tags {
				candidate := append(append([]schemametadata.IMetadataTypeTag{}, row...), filtered...)
				if MetadataTypeTagFactory.Validate(struct {
					Report func(struct {
						Property *string
						Message  string
					}) bool
					Type string
					Tags []schemametadata.IMetadataTypeTag
				}{Report: validateReport, Type: key, Tags: candidate}) {
					rows = append(rows, row)
				}
			}
			atomic.Tags = rows
			if len(atomic.Tags) == 0 {
				atomic.Tags = append(atomic.Tags, filtered)
			} else {
				for i := range atomic.Tags {
					atomic.Tags[i] = append(atomic.Tags[i], filtered...)
				}
			}
		}
	}

	if len(messages) != 0 && props.Errors != nil {
		*props.Errors = append(*props.Errors, MetadataFactory_IError{
			Name:     "comment tag(s)",
			Explore:  props.Explore,
			Messages: messages,
		})
	}
}

func metadataCommentTagFactory_parse(props struct {
	Report func(msg string) any
	Tag    schemametadata.IJsDocTagInfo
}) metadataCommentTagFactory_TagRecord {
	next := metadataCommentTagFactory_PARSER[props.Tag.Name]
	if next == nil {
		return metadataCommentTagFactory_TagRecord{}
	}
	value := ""
	if len(props.Tag.Text) != 0 {
		value = props.Tag.Text[0].Text
	}
	if value == "" && props.Tag.Name != "uniqueItems" {
		props.Report("no comment tag value")
		return nil
	}
	return next(struct {
		Report func(msg string) any
		Value  string
	}{Report: props.Report, Value: value})
}

func (metadataCommentTagFactoryNamespace) Get(props struct {
	Kind  string
	Type  string
	Value string
}) []schemametadata.IMetadataTypeTag {
	parser := metadataCommentTagFactory_PARSER[props.Kind]
	var output []schemametadata.IMetadataTypeTag
	if parser != nil {
		output = parser(struct {
			Report func(msg string) any
			Value  string
		}{
			Report: func(msg string) any { return nil },
			Value:  props.Value,
		})[props.Type]
	}
	if output == nil {
		panic("no tag found for (kind: " + props.Kind + ", type: " + props.Type + ").")
	}
	return output
}

var metadataCommentTagFactory_PARSER = map[string]metadataCommentTagFactory_parser{
	"items": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		value := metadataCommentTagFactory_parse_integer(struct {
			Report   func(msg string) any
			Unsigned bool
			Value    string
		}{Report: props.Report, Value: props.Value, Unsigned: true})
		return metadataCommentTagFactory_TagRecord{"array": {
			{Name: "MinItems<" + props.Value + ">", Target: "array", Kind: "minItems", Value: value, Validate: props.Value + " <= $input.length", Exclusive: true, Schema: map[string]any{"minItems": value}},
			{Name: "MaxItems<" + props.Value + ">", Target: "array", Kind: "maxItems", Value: value, Validate: "$input.length <= " + props.Value, Exclusive: true, Schema: map[string]any{"maxItems": value}},
		}}
	},
	"minItems": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		value := metadataCommentTagFactory_parse_integer(struct {
			Report   func(msg string) any
			Unsigned bool
			Value    string
		}{Report: props.Report, Value: props.Value, Unsigned: true})
		return metadataCommentTagFactory_TagRecord{"array": {{Name: "MinItems<" + props.Value + ">", Target: "array", Kind: "minItems", Value: value, Validate: props.Value + " <= $input.length", Exclusive: true, Schema: map[string]any{"minItems": value}}}}
	},
	"maxItems": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		value := metadataCommentTagFactory_parse_integer(struct {
			Report   func(msg string) any
			Unsigned bool
			Value    string
		}{Report: props.Report, Value: props.Value, Unsigned: true})
		return metadataCommentTagFactory_TagRecord{"array": {{Name: "MaxItems<" + props.Value + ">", Target: "array", Kind: "maxItems", Value: value, Validate: "$input.length <= " + props.Value, Exclusive: true, Schema: map[string]any{"maxItems": value}}}}
	},
	"uniqueItems": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_TagRecord{"array": {{Name: "UniqueItems", Target: "array", Kind: "uniqueItems", Value: true, Validate: "$input.length <= 1 || (new Set($input).size === $input.length)", Exclusive: true, Schema: map[string]any{"uniqueItems": true}}}}
	},
	"type": metadataCommentTagFactory_parse_type,
	"minimum": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_numeric(props, "Minimum", "minimum", props.Value+" <= $input", props.Value+" <= $input", []string{"minimum", "exclusiveMinimum"})
	},
	"maximum": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_numeric(props, "Maximum", "maximum", "$input <= "+props.Value, "$input <= "+props.Value, []string{"maximum", "exclusiveMaximum"})
	},
	"exclusiveMinimum": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_numeric(props, "ExclusiveMinimum", "exclusiveMinimum", props.Value+" < $input", props.Value+" < $input", []string{"minimum", "exclusiveMinimum"})
	},
	"exclusiveMaximum": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_numeric(props, "ExclusiveMaximum", "exclusiveMaximum", "$input < "+props.Value, "$input < "+props.Value, []string{"maximum", "exclusiveMaximum"})
	},
	"multipleOf": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_numeric(props, "MultipleOf", "multipleOf", "$input % "+props.Value+" === 0", "$input % "+props.Value+"n === 0n", true)
	},
	"format": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		matched, ok := metadataCommentTagFactory_FORMATS[props.Value]
		if ok == false {
			return metadataCommentTagFactory_TagRecord{}
		}
		return metadataCommentTagFactory_TagRecord{"string": {{Name: "Format<" + strconv.Quote(matched[0]) + ">", Target: "string", Kind: "format", Value: matched[0], Validate: matched[1], Exclusive: true, Schema: map[string]any{"format": matched[0]}}}}
	},
	"pattern": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		return metadataCommentTagFactory_TagRecord{"string": {{Name: "Pattern<" + strconv.Quote(props.Value) + ">", Target: "string", Kind: "pattern", Value: props.Value, Validate: "RegExp(" + strconv.Quote(props.Value) + ").test($input)", Exclusive: []string{"format"}, Schema: map[string]any{"pattern": props.Value}}}}
	},
	"length": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		value := metadataCommentTagFactory_parse_number(props)
		return metadataCommentTagFactory_TagRecord{"string": {
			{Name: "MinLength<" + props.Value + ">", Target: "string", Kind: "minLength", Value: value, Validate: props.Value + " <= $input.length", Exclusive: true, Schema: map[string]any{"minLength": value}},
			{Name: "MaxLength<" + props.Value + ">", Target: "string", Kind: "maxLength", Value: value, Validate: "$input.length <= " + props.Value, Exclusive: true, Schema: map[string]any{"maxLength": value}},
		}}
	},
	"minLength": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		value := metadataCommentTagFactory_parse_number(props)
		return metadataCommentTagFactory_TagRecord{"string": {{Name: "MinLength<" + props.Value + ">", Target: "string", Kind: "minLength", Value: value, Validate: props.Value + " <= $input.length", Exclusive: true, Schema: map[string]any{"minLength": value}}}}
	},
	"maxLength": func(props struct {
		Report func(msg string) any
		Value  string
	}) metadataCommentTagFactory_TagRecord {
		value := metadataCommentTagFactory_parse_number(props)
		return metadataCommentTagFactory_TagRecord{"string": {{Name: "MaxLength<" + props.Value + ">", Target: "string", Kind: "maxLength", Value: value, Validate: "$input.length <= " + props.Value, Exclusive: true, Schema: map[string]any{"maxLength": value}}}}
	},
}

func metadataCommentTagFactory_parse_type(props struct {
	Report func(msg string) any
	Value  string
}) metadataCommentTagFactory_TagRecord {
	value := props.Value
	if strings.HasPrefix(value, "{") && strings.HasSuffix(value, "}") {
		value = value[1 : len(value)-1]
	}
	if value == "int" {
		value = "int32"
	} else if value == "uint" {
		value = "uint32"
	}
	if metadataCommentTagFactory_includes([]string{"int32", "uint32", "int64", "uint64", "float", "double"}, value) == false {
		return metadataCommentTagFactory_TagRecord{}
	}
	validate := "true"
	if value == "int32" {
		validate = "Math.floor($input) === $input && -2147483648 <= $input && $input <= 2147483647"
	} else if value == "uint32" {
		validate = "Math.floor($input) === $input && 0 <= $input && $input <= 4294967295"
	} else if value == "int64" {
		validate = "Math.floor($input) === $input && -9223372036854775808 <= $input && $input <= 9223372036854775807"
	} else if value == "uint64" {
		validate = "Math.floor($input) === $input && 0 <= $input && $input <= 18446744073709551615"
	} else if value == "float" {
		validate = "-1.175494351e38 <= $input && $input <= 3.4028235e38"
	}
	var numberSchema any
	if value == "int32" || value == "int64" {
		numberSchema = map[string]any{"type": "integer"}
	} else if value == "uint32" || value == "uint64" {
		numberSchema = map[string]any{"type": "integer", "minimum": 0}
	}
	var bigintSchema any
	if value == "uint64" {
		bigintSchema = map[string]any{"minimum": 0}
	}
	return metadataCommentTagFactory_TagRecord{
		"number": {{Name: "Type<" + strconv.Quote(value) + ">", Target: "number", Kind: "type", Value: value, Validate: validate, Exclusive: true, Schema: numberSchema}},
		"bigint": {{Name: "Type<" + strconv.Quote(value) + ">", Target: "bigint", Kind: "type", Value: value, Validate: map[bool]string{true: "true", false: "BigInt(0) <= $input"}[value == "int64"], Exclusive: true, Schema: bigintSchema}},
	}
}

func metadataCommentTagFactory_numeric(props struct {
	Report func(msg string) any
	Value  string
}, name string, kind string, numberValidate string, bigintValidate string, exclusive any) metadataCommentTagFactory_TagRecord {
	number := metadataCommentTagFactory_parse_number(props)
	integer := metadataCommentTagFactory_parse_integer(struct {
		Report   func(msg string) any
		Unsigned bool
		Value    string
	}{Report: props.Report, Value: props.Value, Unsigned: false})
	var bigint any
	if integer != nil {
		bigint = int64(*integer)
	}
	return metadataCommentTagFactory_TagRecord{
		"number": {{Name: name + "<" + props.Value + ">", Target: "number", Kind: kind, Value: number, Validate: numberValidate, Exclusive: exclusive, Schema: map[string]any{kind: number}}},
		"bigint": {{Name: name + "<" + props.Value + "n>", Target: "bigint", Kind: kind, Value: bigint, Validate: bigintValidate, Exclusive: exclusive, Schema: map[string]any{kind: number}}},
	}
}

func metadataCommentTagFactory_parse_number(props struct {
	Report func(msg string) any
	Value  string
}) any {
	parsed, err := strconv.ParseFloat(props.Value, 64)
	if err != nil || math.IsNaN(parsed) {
		props.Report("invalid number")
		return nil
	}
	return parsed
}

func metadataCommentTagFactory_parse_integer(props struct {
	Report   func(msg string) any
	Unsigned bool
	Value    string
}) *int {
	parsedAny := metadataCommentTagFactory_parse_number(struct {
		Report func(msg string) any
		Value  string
	}{Report: props.Report, Value: props.Value})
	parsed, ok := parsedAny.(float64)
	if ok == false {
		return nil
	}
	if math.Floor(parsed) != parsed {
		props.Report("invalid integer")
		return nil
	}
	if props.Unsigned && parsed < 0 {
		props.Report("invalid unsigned integer")
		return nil
	}
	value := int(parsed)
	return &value
}

var metadataCommentTagFactory_FORMATS = func() map[string][2]string {
	output := map[string][2]string{}
	for key, value := range FormatCheatSheet {
		output[key] = [2]string{key, value}
	}
	output["datetime"] = [2]string{"date-time", "!isNaN(new Date($input).getTime())"}
	output["dateTime"] = [2]string{"date-time", "!isNaN(new Date($input).getTime())"}
	return output
}()

func metadataCommentTagFactory_includes(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}
